Instrument View#render instead of DispatcherServlet#render

kananindzya opened the pull request:
<!--

Replace this comment with a description of what's being changed by this PR.

If this PR should close an issue, please add one of the magic keywords
(e.g. Fixes) followed by the issue number. For more info see:
https://help.github.com/articles/closing-issues-using-keywords/

If this pull request is work in progress, create a draft PR instead of prefixing the title with WIP.

-->

### Checklist

<!-- Potential tasks related to a new PR. Remove tasks that are not relevant -->
closes elastic/apm-agent-java#811 
- [x] Implement code
- [x] Add tests
-- [x] jsp
-- [x] markup templates
-- [x] freemarker
-- [x] groovy templates
-- [x] jackson2json view
-- [x] jade4j template
- [ ] Update documentation
- [x] Update [CHANGELOG.md](CHANGELOG.md)
- [x] Update [supported-technologies.asciidoc](docs/supported-technologies.asciidoc)
- [ ] Added an API method or config option? Document in which version this will be introduced.

kananindzya commented:
Hi @felixbarny ,
please review pull request.

kananindzya commented:
![image](https://user-images.githubusercontent.com/48118512/64225963-6223b380-ceff-11e9-8dec-65641c475843.png)
![image](https://user-images.githubusercontent.com/48118512/64225984-7b2c6480-ceff-11e9-8485-93e012df87c2.png)

felixbarny commented on apm-agent-plugins/apm-spring-webmvc-plugin/src/main/java/co/elastic/apm/agent/spring/webmvc/ViewRenderInstrumentation.java:
```diff
@@ -0,0 +1,123 @@
+            }
+            final TraceContextHolder<?> parent = tracer.getActive();
+
+            String viewClassName = thiz.getClass().getSimpleName().replace("View", "");
```
We can be a bit more efficient here. 

Similar to this (I have not tested it)

```java
StringBuilder name = span.getAndOverrideName(PRIO_DEFAULT);
if (name != null) {
    String className = thiz.getClass().getName();
    name.append(className, className.lastIndexOf('.'), className.lastIndexOf("View"));
}
```

kananindzya replied:
replaced on:
```
 String className = thiz.getClass().getName();
            span = parent.createSpan()
                .withType(SPAN_TYPE)
                .withSubtype(className.substring(className.lastIndexOf('.')+1, className.lastIndexOf("View")))
                .withAction(SPAN_ACTION)
                .withName(DISPATCHER_SERVLET_RENDER_METHOD);

```

felixbarny replied:
Substring still creates a new object. With my suggestion, there would be no allocation involved as it just copies over a range of characters from the class name string to the string builder.

felixbarny replied:
oh, right 🤦‍♂️
nevermind then

felixbarny commented on apm-agent-plugins/apm-spring-webmvc-plugin/src/main/java/co/elastic/apm/agent/spring/webmvc/ViewRenderInstrumentation.java:
```diff
@@ -0,0 +1,123 @@
+            }
+            switch (viewClassName) {
+                case "Thymeleaf":
+                    AbstractThymeleafView view = (AbstractThymeleafView) thiz;
```
This may cause linkage errors if `AbstractThymeleafView` is not on the class path. Better to invoke it via reflection. It's also ok to just not support `AbstractThymeleafView` in the first version. We can always add that later.

kananindzya replied:
removed

felixbarny commented on apm-agent-plugins/apm-spring-webmvc-plugin/src/main/java/co/elastic/apm/agent/spring/webmvc/ViewRenderInstrumentation.java:
```diff
@@ -0,0 +1,123 @@
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Arrays.asList("thymeleaf-view", "freemarker-view", "jackson2json-view", "jsp-view", "groovy-template-view", "jade4j-template-view");
```
```suggestion
        return Arrays.asList("spring-view-render");
```

kananindzya replied:
changed

felixbarny commented on apm-agent-plugins/apm-spring-webmvc-plugin/src/main/java/co/elastic/apm/agent/spring/webmvc/ViewRenderInstrumentation.java:
```diff
@@ -0,0 +1,123 @@
+            }
+        }
+    }
+
```
add a `getTypeMatcherPreFilter` with `nameContains("View")` to speed up the matching 🙂

kananindzya replied:
added

kananindzya commented on apm-agent-plugins/apm-spring-webmvc-plugin/src/main/java/co/elastic/apm/agent/spring/webmvc/ViewRenderInstrumentation.java:
```diff
@@ -0,0 +1,117 @@
+            String className = thiz.getClass().getName();
+            span = parent.createSpan()
+                .withType(SPAN_TYPE)
+                .withSubtype(className.substring(className.lastIndexOf('.')+1, className.lastIndexOf("View")))
```
@eyalkoren 
Should I add this method: 
```
        public static String defineSubtype(String className) {
            switch (className) {
                case "org.springframework.web.servlet.view.groovy.GroovyMarkupView":
                    return "GroovyMarkup";
                case "org.springframework.web.servlet.view.freemarker.FreeMarkerView":
                    return "FreeMarker";
                case "org.springframework.web.servlet.view.json.MappingJackson2JsonView":
                    return "MappingJackson2Json";
                case "de.neuland.jade4j.spring.view.JadeView":
                    return "Jade";
                case "org.springframework.web.servlet.view.InternalResourceView":
                    return "InternalResource";
                case "org.thymeleaf.spring4.view.ThymeleafView":
                    return "Thymeleaf";
                default:
                    break;
            }
            return className.substring(className.lastIndexOf('.')+1, className.lastIndexOf("View"));
        }
```
?

kananindzya replied:
added

eyalkoren commented on apm-agent-plugins/apm-spring-webmvc-plugin/src/main/java/co/elastic/apm/agent/spring/webmvc/ViewRenderInstrumentation.java:
```diff
@@ -0,0 +1,143 @@
+            }
+        }
+
+        public static String defineSubtype(String className) {
```
```suggestion
        @VisibleForAdvice
        public static String obtainSubtype(String className) {
```

eyalkoren commented on apm-agent-plugins/apm-spring-webmvc-plugin/src/main/java/co/elastic/apm/agent/spring/webmvc/ViewRenderInstrumentation.java:
```diff
@@ -106,9 +106,14 @@ public static String defineSubtype(String className) {
                     return "Thymeleaf";
                 default:
-                    break;
+                    if (subTypeCache.containsKey(className)) {
```
Usually it's better get and check if `null`, as in:
```
String subType = subTypeCache.get(className);
if (subType == null) {
    subType = className.substring(className.lastIndexOf('.') + 1, className.lastIndexOf("View"));
    if (subTypeCache.size() < 1000) {
        subTypeCache.put(className, subType);
    }
}
return sunType;
```
Then it's one lookup in the cache instead of two every time

eyalkoren commented on apm-agent-plugins/apm-spring-webmvc-plugin/src/main/java/co/elastic/apm/agent/spring/webmvc/ViewRenderInstrumentation.java:
```diff
@@ -0,0 +1,143 @@
+                        return subTypeCache.get(className);
+                    } else {
+                        String subType = className.substring(className.lastIndexOf('.') + 1, className.lastIndexOf("View"));
+                        subTypeCache.put(className, subType);
```
```suggestion
                        if (subTypeCache.size() < 1000) {
                            subTypeCache.put(className, subType);
                        }
```
Just precaution from strange unknown scenarios

felixbarny commented:
I've added the suggestions, hope you don't mind 🙂 
Also found and fixed some edge cases which could lead to a `StringIndexOutOfBoundsException` for really strange class names and for those not containing `"View"`.

kananindzya commented:
> I've added the suggestions, hope you don't mind slightly_smiling_face
> Also found and fixed some edge cases which could lead to a `StringIndexOutOfBoundsException` for really strange class names and for those not containing `"View"`.

Thanks :) 
I also thought about classes without packages

felixbarny commented:
Thanks for bearing with us even if it sometimes takes some time to get changes in, but we really appreciate it!
